### PR TITLE
flag_exists.cmd to cexec

### PR DESCRIPTION
**Pull request by daxgames**
Updates to https://github.com/cmderdev/cmder/pull/1758

`flag_exists.cmd` does not adequately describe what this does.  It is actually a conditional execution based on a detected/undetected flag passed to `init.bat`.

In my PR:

* Renamed `%cmder_root%/vendor/lib/flag_exists.cmd` to `%cmder_root%/vendor/bin/cexec.cmd`
   * This makes the file easily callable by its name since `%cmder_root%/vendor/bin/` is added to the path
* Modified the `cexec /setpath` to set:
   * `ccall=call C:\Users\user\cmderdev\vendor\bin\cexec.cmd` Evaluates flags, runs commands if found,  and returns to the calling script and continues.
      * Example: `%ccall% /startnotepad start notepad.exe`
   * `cexec=C:\Users\user\cmderdev\vendor\bin\cexec.cmd` Evaluates flags, runs commands if found, and does not return to the calling script.
      * Example: `%cexec% /startnotepad start notepad.exe`
* Added exit codes to allow multi conditional execution:
   * If a flag is detected and the command is run `cexec` exists with an `%ERRORLEVEL%=0`
   * If `not` is specified and a flag is not detected and the command is run `cexec` exists with an `%ERRORLEVEL%=0`
   * Else `cexec` exists with `%ERRORLEVEL%=1`
   * Example: `%ccall% "/startNotepad" "start" "notepad.exe" && %call% not "/startwordpad" "start" "wordpad.exe"`
      * Wordpad is only started if `%CMDER_USER_FLAGS%` contains both `/startNotepad and /startwordpad`
* Enable ability to conditionally set environment variables in the calling session.

`user_profile.cmd`

```
%ccall% "/editorvim" "set" "EDITOR=vim"
echo EDITOR=%EDITOR%
```

* Updated the documentation in the `readme.md`.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -148,7 +148,7 @@ You may find some Monokai color schemes for mintty to match Cmder [here](https:/
 | `/svn_ssh [path to ssh.exe]`    | Define `%SVN_SSH%` so we can use git svn with ssh svn repositories.                              | `%GIT_INSTALL_ROOT%\bin\ssh.exe`      |
 | `/user_aliases [file path]`     | File path pointing to user aliases.                                                              | `%CMDER_ROOT%\config\user-liases.cmd` |
 | `/v`                            | Enables verbose output.                                                                          | not set                               |
-| (custom arguments)              | User defined arguments processed by `flag_exists`. Type `%flag_exists% /?` for more useage.      | not set                               |
+| (custom arguments)              | User defined arguments processed by `cexec`. Type `cexec /?` for more useage.                    | not set                               |
```
Comment (xiazeyu) [inline README.md]: Conflict with below?
https://github.com/cmderdev/cmder/pull/1758#issuecomment-423712248

File: README.md [hunk 2]
```diff
@@ -247,39 +247,53 @@ Uncomment and edit the below line in the script to use Cmder config even when la
...
 
-```batch
-
-init.bat /startNotepad
-
-```
-
-to start init.bat with custom argument(`/startNotepad`) and put
+  ```
+  ccall=call C:\Users\user\cmderdev\vendor\bin\cexec.cmd
```
Comment (xiazeyu) [inline README.md]: Is it necessary to be in the usage introduction?

File: README.md [hunk 3]
```diff
@@ -247,39 +247,53 @@ Uncomment and edit the below line in the script to use Cmder config even when la
...
+  ```  
 
-```batch
+  Example: `%ccall% /startnotepad start notepad.exe`
```
Comment (xiazeyu) [inline README.md]: https://github.com/cmderdev/cmder/pull/1758#issuecomment-423712248

File: README.md [hunk 4]
```diff
@@ -247,39 +247,53 @@ Uncomment and edit the below line in the script to use Cmder config even when la
...
+  
+  ```
 
-```
+* Add the below to your `%cmder_root%\config\user_profile.cmd`
 
-the `notepad.exe` won't be executed.
+  ```batch
+  
+  %ccall% "/startNotepad" "start" "notepad.exe"`
```
Comment (xiazeyu) [inline README.md]: https://github.com/cmderdev/cmder/pull/1758#issuecomment-423712248

File: vendor/bin/cexec.cmd [hunk 5]
```diff
@@ -19,7 +19,8 @@ set "currenArgu=%~1"
 if /i "%currenArgu%" equ "/setPath" (
```
Comment (xiazeyu) [inline vendor/bin/cexec.cmd]: https://github.com/cmderdev/cmder/pull/1758#issuecomment-423712248

File: vendor/bin/cexec.cmd [hunk 6]
```diff
@@ -104,24 +107,25 @@ echo                    to wrap your flag name to avoid exceed expectation.
...
+echo   These examples are expected to be written in %cmder_root%/config/user-profile.cmd
+echo   CExec evaluates the environment varible "CMDER_USER_FLAGS" and conditionally
+echo   caries out actions based on flags that are passed.
 echo.
 echo   Case 1:
 echo.
 echo   The following command in user-profile.cmd would execute "notepad.exe"
 echo.
-echo     call %%flag_exists%% "/startNotepad" "start" "notepad.exe"
+echo     call cexec "/startNotepad" "start" "notepad.exe"
```
Comment (xiazeyu) [inline vendor/bin/cexec.cmd]: `ccall "/startNotepad" "start" "notepad.exe"`?

File: README.md (hunk 1 above)
Comment (daxgames) [reply README.md]: I don't understand the comment?

File: vendor/bin/cexec.cmd (hunk 6 above)
Comment (daxgames) [reply vendor/bin/cexec.cmd]: It will work either way `%ccall%` is a shortcut for `call cexec`

Comment (xiazeyu) [reply vendor/bin/cexec.cmd]: usage of %ccall% didn't appear in cexec, and I think there's a confusion among `%ccall%`, `ccall`, `%cexec%` and `cexec`. I think only `cexec` is documented.

**Comment by daxgames** [comment]
Either way you can run it two ways so I gave it two shortcuts.

`call cexec` = `%ccall%`
`cexec` = `%cexec%`

**Comment by daxgames** [comment]
I will look at it tonight.  I thought I documented it pretty clearly, I guess I still don't understand your concern but I will look closer any try to clarify.

**Comment by daxgames** [comment]
Below is the documentation of cexec as it exists in the `README.md` this PR. 

--

### Customizing user sessions using `init.bat` custom arguments. 

You can pass custom arguments to `init.bat` and use `cexec.cmd` in your `user_profile.cmd` to evaluate these
arguments then execute commands based on a particular flag being detected or not.

`init.bat` creates two shortcuts for using `cexec.cmd` in your profile scripts.

#### `%ccall%` - Evaluates flags, runs commands if found,  and returns to the calling script and continues.

```
ccall=call C:\Users\user\cmderdev\vendor\bin\cexec.cmd
```  

Example: `%ccall% /startnotepad start notepad.exe`

#### `%cexec%` - Evaluates flags, runs commands if found, and does not return to the calling script.

```
cexec=C:\Users\user\cmderdev\vendor\bin\cexec.cmd
```

Example: `%cexec% /startnotepad start notepad.exe`

It is useful when you have multiple tasks to execute `cmder` and need it to initialize
the session differently depending on the task chosen.

To conditionally start `notepad.exe` when you start a specific `cmder` task:

* Press <kbd>win</kbd>+<kbd>alt</kbd>+<kbd>t</kbd>
* Click `+` to add a new task.
* Add the below to the `Commands` block:

  ```batch
  
  cmd.exe /k ""%ConEmuDir%\..\init.bat" /startnotepad"
  
  ```

* Add the below to your `%cmder_root%\config\user_profile.cmd`

  ```batch
  
  %ccall% "/startNotepad" "start" "notepad.exe"`
  
  ```

To see detailed usage of `cexec`, type `cexec /?` in cmder.

--

It clearly documents both the `%ccall%` and the `%cexec%` shortcuts and the differences between them.  I, like you, want the documentation to be clear and helpful and I think it is.  What exactly do you think needs to change?

Below is the command line help for `cexec.cmd`:

--

```
echo.
echo CExec - Conditional Exec
echo.
echo Handles with custom arguments for cmder's init.bat.
echo   written by xiazeyu, inspired DRSDavidSoft.
echo.
echo Usage:
echo.
echo cexec /setPath [NOT] flagName command/program [parameters]
echo.
echo   /setPath         Generate a global varibles %%ccall%% and  %%cexec%% for
echo                    quicker use. Following arguments will be ignored.
echo.
echo   NOT              Specifies that cexec should carry out
echo                    the command only if the flag is missing.
echo.
echo   /[flagName]      Specifies which flag name is to detect. It's recommand
echo                    to use a pair of double quotation marks to wrap
echo                    your flag name to avoid exceed expectation.
echo.
echo   command/program  Specifies the command to carry out if the
echo                    argument name is detected. It's recommand to
echo                    use a pair of double quotation marks to
echo                    wrap your command to avoid exceed expectation.
echo.
echo   parameters       These are the parameters passed to the command/program.
echo                    It's recommand to use a pair of double quotation marks 
echo                    to wrap your flag name to avoid exceed expectation.
echo.
echo Examples:
echo.
echo   These examples are expected to be written in %cmder_root%/config/user-profile.cmd
echo   CExec evaluates the environment varible "CMDER_USER_FLAGS" and conditionally
echo   caries out actions based on flags that are passed.
echo.
echo   Case 1:
echo.
echo   The following command in `user_profile.cmd` would execute "notepad.exe" and continue running the `user_profile.cmd`
echo.
echo     "%ccall%" "/startNotepad" "start" "notepad.exe"
echo.
echo   If you pass parameter to init.bat like:
echo.
echo     init.bat /startNotepad
echo.
echo   Case 2:
echo.
echo   The following command in `user_profile.cmd` would execute "notepad.exe" and stop running the `user_profile.cmd`
echo.
echo     "%cexec%" NOT "/dontStartNotepad" "start" "notepad.exe"
echo.
echo   UNLESS you pass parameter to init.bat like:
echo.
echo     init.bat /dontStartNotepad
echo.
```

I made a couple of changes and hopefully it is more to your liking.

Please let me know.

**Pull request merged**